### PR TITLE
feat(hostRules/matchHost): massage and validate 

### LINKED ISSUE #29425: hostRules: improve matchHost validation and parsing
### Describe the proposed change(s).

Goal: massage "incorrect" matchHost values with ports or paths, for example `gitlab.company.com:8080` or `registry.company.com/some/path`.

Improve validation: if a matchHost value contains `"` or `/` then we should add `https://` as the prefix and check if it's a valid URL. If it's not then I think it's safe for us to raise a config validation error. If it's valid, then we don't need to migrate it, because some config.js configurations may not be possible to migrate (the value comes from env).

Also improve validation of empty matchHost. If the `matchHost` key is defined, but the _value_ is `undefined`, then we should warn. This is commonly caused by empty process.env.SOMETHING being used. We should similar do for empty string `""`.

Finally, we should enhance the addRule function to do similar logic and prefix the matchHost with `https://` if it contains `:` or `/`.

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Prefix `matchHost` url which contain `:` or `/` with `htps://` (migrated as well)
- Validate `matchHost` urls
   1. if its an empty string
   2. check is it is a valid url if the url includes `://` (ie. only check for valid if url has scheme)
   
   

## Context
Closes: https://github.com/renovatebot/renovate/issues/29425

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/validation.spec.ts [hunk 1]
```diff
@@ -1134,6 +1134,44 @@ describe('config/validation', () => {
       ]);
     });
 
+    it('errors if invalid matchHost values in hostRules', async () => {
+      GlobalConfig.set({ allowedHeaders: ['X-*'] });
+
+      const config = {
+        hostRules: [
```
Comment (rarkins) [inline lib/config/validation.spec.ts:1141]: Please add one with a missing matchHost to make sure that _doesn't_ error. e.g. have hostType and token only

File: lib/config/migrations/custom/host-rules-migration.spec.ts [hunk 2]
```diff
@@ -27,6 +27,7 @@ describe('config/migrations/custom/host-rules-migration', () => {
           { hostName: 'some.domain.com', token: '123test' },
           { endpoint: 'domain.com/', token: '123test' },
           { host: 'some.domain.com', token: '123test' },
+          { host: 'some.domain.com:8080', token: '123test' },
```
Comment (rarkins) [inline lib/config/migrations/custom/host-rules-migration.spec.ts:30]:
```suggestion
          { matchHost: 'some.domain.com:8080', token: '123test' },
```

File: lib/config/validation.ts [hunk 3]
```diff
@@ -814,6 +814,25 @@ export async function validateConfig(
...
+        if (is.nonEmptyString(rule.matchHost)) {
+          if (rule.matchHost.includes('://')) {
+            try {
+              new URL(rule.matchHost);
+            } catch (err) {
+              errors.push({
+                topic: 'Configuration Error',
+                message: `hostRules matchHost \`${rule.matchHost}\` is not a valid URL.`,
+              });
+            }
```
Comment (rarkins) [inline lib/config/validation.ts:826]: Use parseUrl() from utils instead, then no need to try/catch

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
Please test for both global and repo config

**Comment by RahulGautamSingh** [comment]
> Please test for both global and repo config

Done ☑️

**Review by rarkins — DISMISSED** [review]

File: lib/config/validation.ts [hunk 4]
```diff
@@ -814,6 +815,23 @@ export async function validateConfig(
           ? (config.allowedHeaders as string[]) ?? []
           : GlobalConfig.get('allowedHeaders', []);
       for (const rule of val as HostRule[]) {
+        if (is.nonEmptyString(rule.matchHost)) {
+          if (rule.matchHost.includes('://')) {
+            if (is.null_(parseUrl(rule.matchHost))) {
```
Comment (viceice) [inline lib/config/validation.ts:820]:
```suggestion
            if (null === parseUrl(rule.matchHost)) {
```
looks simpler than the `null_` function 🤔

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/config/validation.ts (hunk 4 above)
Comment (RahulGautamSingh) [inline lib/config/validation.ts:820]:
```suggestion
            if (parseUrl(rule.matchHost) === null) {
```

File: lib/config/migrations/custom/host-rules-migration.ts [hunk 5]
```diff
@@ -94,6 +94,8 @@ function validateHostRule(rule: LegacyHostRule & HostRule): void {
 function massageUrl(url: string): string {
```
Comment (viceice) [inline lib/config/migrations/custom/host-rules-migration.ts:94]: we should use a single shared function here and inside host rules

**Review by viceice — COMMENTED** [review]
otherwise LGTM

**Comment by RahulGautamSingh** [comment]
When I export the `massageHostUrl` fn from `util/host-rules` and import it in `host-rules-migration.ts` the `util/sanitize` module doesn't get mocked properly in the `datasource/npm/npmrc.spec.ts` file. I am not sure what is the cause & the fix here.

If I only export the `massageHostUrl` fn from `util/host-rules` but do not import it in `host-rules-migration.ts` the `util/sanitize` module is mocked.

@viceice any suggestions?

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**